Add WKKeyedCoder as an alternate code path for WebKit secure coding types (and use it to serialize AVOutputContext)

beidson opened the pull request:
#### 361273dff3a1320574cb743de6930e21cbea87a6
<pre>
Add WKKeyedCoder as an alternate code path for WebKit secure coding types (and use it to serialize AVOutputContext)
<a href="https://bugs.webkit.org/show_bug.cgi?id=265972">https://bugs.webkit.org/show_bug.cgi?id=265972</a>
<a href="https://rdar.apple.com/119279237">rdar://119279237</a>

Reviewed by Alex Christensen.

For some types where adopting the WebKit property list format is slow-to-happen or not possible
(e.g. older operating systems) we can gather the equivalent property list ourselves with our own NSCoder.

WKKeyedCoder fills that role. It accumulates the key/value pairs encoded into an NSDictionary which
is then treated by the CoreIPC machinery as equivalent to a WebKit secure coding property list.

On the decode side, once validated like other WebKit property lists, it is used to recreate the object.

I chose a very simple first class to use this feature on - AVOutputContext and its (2) NSString members.
As we apply this strategy to more advanced classes we will also flesh out WKKeyedCoder&apos;s feature set.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(generate_webkit_secure_coding_impl):
(generate_webkit_secure_coding_impl.is):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp:
(WebKit::dictionaryForWebKitSecureCodingType):
(WebKit::CoreIPCAVOutputContext::CoreIPCAVOutputContext):
(WebKit::CoreIPCAVOutputContext::isValidDictionary):
(WebKit::CoreIPCAVOutputContext::toID const):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h:
(WebKit::CoreIPCAVOutputContext::CoreIPCAVOutputContext):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;WebKit::CoreIPCAVOutputContext&gt;::encode):
(IPC::ArgumentCoder&lt;WebKit::CoreIPCAVOutputContext&gt;::decode):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass&lt;AVOutputContext&gt;):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCAVOutputContext.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/WKKeyedCoder.h: Added.
* Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm: Added.
(-[WKKeyedCoder init]):
(-[WKKeyedCoder initWithDictionary:]):
(-[WKKeyedCoder allowsKeyedCoding]):
(-[WKKeyedCoder encodeObject:forKey:]):
(-[WKKeyedCoder containsValueForKey:]):
(-[WKKeyedCoder decodeObjectOfClass:forKey:]):
(-[WKKeyedCoder decodeObjectForKey:]):
(-[WKKeyedCoder accumulatedDictionary]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/271664@main">https://commits.webkit.org/271664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fce7a2689d9c56a047599946ac9391e0a28cc45e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29170 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7839 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30506 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31757 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26529 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5144 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26534 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29442 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24990 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5607 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29323 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33094 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26459 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3892 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29757 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7387 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6961 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6223 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -1444,14 +1448,28 @@ def generate_webkit_secure_coding_impl(serialized_types, headers):
     result.append('')
     result.append('namespace WebKit {')
     result.append('')
+    result.append('static NSDictionary *dictionaryForWebKitSecureCodingType(id object)')
```
We could make this return a RetainPtr<NSDictionary> so accumulatedDictionary doesn't need to add to the autorelease pool.

beidson replied:
Done

achristensen07 commented on Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm:
```diff
@@ -0,0 +1,103 @@
+
+@implementation WKKeyedCoder {
+    RetainPtr<NSMutableDictionary> m_mutableDictionary;
+    RetainPtr<NSDictionary> m_dictionary;
```
This seems error prone.  Can't we just have one dictionary and have it be mutable?

beidson replied:
Done

achristensen07 commented on Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
```diff
@@ -638,6 +642,13 @@ String certDataBase64(""
+    // AVOutputContext
+#if USE(AVFOUNDATION)
+    RetainPtr<AVOutputContext> outputContext = adoptNS([[PAL::getAVOutputContextClass() alloc] init]);
+    runTestNS({ outputContext.get() });
```
Can we check one with actual real context ID and context type?

beidson replied:
Due to how the class works, this one *does* have a real ID and context type. e.g. "the default AVOutputContext"
`(lldb) po m_dictionary.get()
{
    AVOutputContextSerializationKeyContextID = "9BF612E3-3845-4059-9AA2-03230A5E8B04";
    AVOutputContextSerializationKeyContextType = AVOutputContextTypeVideo;
}`

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCAVOutputContext.serialization.in:
```diff
@@ -0,0 +1,29 @@
+
+#if USE(AVFOUNDATION)
+secure_coding_header: <pal/cocoa/AVFoundationSoftLink.h>
+[WebKitSecureCodingClass=PAL::getAVOutputContextClass(), SupportWKKeyedCoder] webkit_secure_coding AVOutputContext {
```
I was hoping for a comment somewhere like // FIXME: Remove SupportWKKeyedCoder once AVOutputContext conforms to WebKit's secure coding protocol, tracked by rdar://NNNNNNNNNN

achristensen07 commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -272,12 +275,23 @@ - (instancetype)initWithCGColor:(CGColorRef)color
 
     // Specific classes handled.
+#if USE(AVFOUNDATION)
+    if (PAL::isAVFoundationFrameworkAvailable() && [object isKindOfClass:[PAL::getAVOutputContextClass() class]])
```
Again, I really don't think the last class does anything.

beidson replied:
Okay, see it now, was blind to it before.
Copied pattern from others.
You're right - cleared them all out

achristensen07 commented on Source/WebKit/Shared/Cocoa/WKKeyedCoder.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#if PLATFORM(COCOA)
+#ifdef __OBJC__
```
I don't think we need PLATFORM(COCOA)

beidson committed:
Add WKKeyedCoder as an alternate code path for WebKit secure coding types (and use it to serialize AVOutputContext)
https://bugs.webkit.org/show_bug.cgi?id=265972
rdar://119279237

Reviewed by Alex Christensen.

For some types where adopting the WebKit property list format is slow-to-happen or not possible
(e.g. older operating systems) we can gather the equivalent property list ourselves with our own NSCoder.

WKKeyedCoder fills that role. It accumulates the key/value pairs encoded into an NSDictionary which
is then treated by the CoreIPC machinery as equivalent to a WebKit secure coding property list.

On the decode side, once validated like other WebKit property lists, it is used to recreate the object.

I chose a very simple first class to use this feature on - AVOutputContext and its (2) NSString members.
As we apply this strategy to more advanced classes we will also flesh out WKKeyedCoder's feature set.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(generate_webkit_secure_coding_impl):
(generate_webkit_secure_coding_impl.is):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp:
(WebKit::dictionaryForWebKitSecureCodingType):
(WebKit::CoreIPCAVOutputContext::CoreIPCAVOutputContext):
(WebKit::CoreIPCAVOutputContext::isValidDictionary):
(WebKit::CoreIPCAVOutputContext::toID const):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h:
(WebKit::CoreIPCAVOutputContext::CoreIPCAVOutputContext):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::ArgumentCoder<WebKit::CoreIPCAVOutputContext>::encode):
(IPC::ArgumentCoder<WebKit::CoreIPCAVOutputContext>::decode):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::getClass<AVOutputContext>):
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCAVOutputContext.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/WKKeyedCoder.h: Added.
* Source/WebKit/Shared/Cocoa/WKKeyedCoder.mm: Added.
(-[WKKeyedCoder init]):
(-[WKKeyedCoder initWithDictionary:]):
(-[WKKeyedCoder allowsKeyedCoding]):
(-[WKKeyedCoder encodeObject:forKey:]):
(-[WKKeyedCoder containsValueForKey:]):
(-[WKKeyedCoder decodeObjectOfClass:forKey:]):
(-[WKKeyedCoder decodeObjectForKey:]):
(-[WKKeyedCoder accumulatedDictionary]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(TEST):

Canonical link: https://commits.webkit.org/271664@main
